runtime/libtapasco/src/allocator.rs

```diff
@@ -0,0 +1,296 @@
+                    self.memory_free[i],
+                    self.memory_free[n]
+                );
+                self.memory_free[i].size += self.memory_free[n].size;
+                self.memory_free.remove(n);
+            } else {
+                trace!("Checking next element.");
+                i += 1;
+            }
+        }
+    }
+}
+
+impl Allocator for GenericAllocator {
+    fn allocate(&mut self, size: DeviceSize) -> Result<DeviceAddress> {
+        if size == 0 {
+            return Err(Error::InvalidSize { size: size });
+        }
+        trace!("Looking for free memory.");
+        let size_aligned = self.fix_alignment(size);
+        let mut element_found = None;
+        let mut addr_found = None;
+        for (i, s) in &mut self.memory_free.iter_mut().enumerate() {
+            if s.size >= size_aligned {
+                trace!("Found free space in segment {:?}.", s);
+                let addr = s.base;
+                addr_found = Some(addr);
+                self.memory_used.push(MemoryFree {
+                    base: addr,
+                    size: size_aligned,
+                });
+                s.size -= size_aligned;
+                s.base += size_aligned;
+                if s.size == 0 {
+                    element_found = Some(i);
+                } else {
+                    trace!("New segment is {:?}.", s);
+                }
+            }
+        }
+
+        match element_found {
+            Some(x) => {
+                trace!("Removing empty segment.");
+                self.memory_free.remove(x);
+            }
+            None => (),
+        };
+
+        match addr_found {
+            Some(x) => Ok(x),
+            None => Err(Error::OutOfMemory { size: size_aligned }),
+        }
+    }
+
+    fn free(&mut self, ptr: DeviceAddress) -> Result<()> {
+        match self.memory_used.iter().position(|&x| x.base == ptr) {
+            Some(x) => {
+                let m = self.memory_used[x];
+                self.memory_used.remove(x);
+                trace!("Freeing memory segment {:?}", m);
+                if self.memory_free.len() == 0 {
+                    trace!("No free memory right now, adding directly.");
+                    self.memory_free.push(m);
+                } else {
+                    match self.memory_free.iter().position(|&x| x.base > m.base) {
+                        Some(x) => {
+                            trace!("Adding memory left of {}.", x);
+                            self.memory_free.insert(x, m);
+                        }
+                        None => {
+                            trace!("Adding memory to the end.");
+                            self.memory_free.push(m)
+                        }
+                    };
+                }
+                self.merge_memory();
+                Ok(())
+            }
+            None => Err(Error::UnknownMemory { ptr: ptr }),
+        }
+    }
+}
+
+#[test]
```

Comment:
Is there a reason theses tests are here (and in examples/Rust/libtapasco_tests)?

Reply:
The tests are used wherever I felt that they are needed and can be implemented easily right now.

I would love to have more tests, but currently the tests are limited to parts of the code that don't interact with the driver. As soon as we have a dummy version of the driver I'd like to change that.

Reply:
I forgot a "not" in my question. I wanted to say "and **not** in examples/Rust/libtapasco_tests?".
Is there a reason these tests are directly in the source file and not with the rest of the tests in libtapasco_tests?

Reply:
Ohh, those are different kind of tests.

The `libtapasco_tests` is used to do some initial testing on the device, but doesn't really test the code itself.

These tests here marked `#[test]` are from the unit testing framework provided by Rust and are run using `cargo test`. They are supposed to be located right next to the code they affect and can be used wherever the programmer feels that such tests are necessary. There are also corresponding `benchmark` versions for micro benchmarks which would be really useful (as soon as we have a dummy driver).

Reply:
Ok